### ISSUE TITLE
Testing: New batch edit controls

### ISSUE THREAD

**Issue by jimsafley** [omeka-s-modules/Mapping#114]
The [batch-copy-coordinates](https://github.com/omeka-s-modules/Mapping/compare/batch-copy-coordinates) branch adds three new controls to the item batch edit page:

- Delete markers: delete Mapping markers
- Copy coordinates to markers: copy coordinates from property values to Mapping markers
- Update markers: update Mapping marker popup content

The "Delete markers" control is a simple checkbox that instructs the batch process to delete all Mapping markers belonging to selected items. To test this, add markers as normal then use this control to delete them. Verify that the markers were deleted.

The "Copy coordinates to markers" control is a compound control used to instruct the batch process where to find the coordinates and how the coordinates are formatted. There are four copy actions:

- Copy from one item property containing both latitude and longitude
- Copy from two item properties, one latitude and the other longitude
- Copy from one media property containing both latitude and longitude
- Copy from two media properties, one latitude and the other longitude

When copying from one property, the user will select the property, the coordinates order, and the coordinates delimiter. When selecting from two properties, the user will select a latitude property and a longitude property. When copying from media, the user will have an additional option to assign the media to the new markers. To test this, add coordinates to several items and media in a similar fashion to below. The batch process should ignore invalid coordinates. Verify that the markers were created and, if set, the media were assigned.

![image](https://github.com/omeka-s-modules/Mapping/assets/141955/423c22d5-92e9-4f67-9da3-2ab61c976b0d)

(Note above that coordinates are present in both the one property and two property configurations.)

The "Update markers" control is a compound control used to instruct the batch process how to update marker popups. This control lets users update the label and assigned media (i.e. image) of the popups. For "Media image" users have the option to unassign the currently assigned media, or to assign the item's primary media to the markers. Users also have the option to copy a property value to the marker's label. The property value can come from one of three sources: the item, the item's primary media, or the marker's assigned media. To test this, assign the primary media to the marker and verify that the popups display the media image and a link to the media. Then, unassign the media and verify that it worked. Then assign a label from every one of the possible sources and verify that the markers have the correct labels.

**Comment by allanaaa** [comment]
Small quibble: Can you reorder the batch-edit options so that "Update markers" and "Delete markers" go together? Maybe it should go Create, Update, Delete, in that order.

**Comment by jimsafley** [comment]
I'm hesitant to reorder because delete happens first.

**Comment by allanaaa** [comment]
Everything works nicely that I can see. Here's what I've drafted for this. 

> ### Batch-edit map data

> Users can select multiple items and perform batch editing to create and edit map markers. The options are:
> 
> - **Delete markers**: You can batch-remove all existing markers from multiple items. 
> - **Copy coordinates to markers**: This involves taking latitude and longitude data from an existing metadata value on each item or from their associated media. This will append new markers, not overwrite any existing markers. If there are multiple values in the chosen proprerty, this will batch-copy them all as multiple markers. You can specify:
> 	- Which fields contain each or both values
> 	- How they are separated (the operation will ignore spaces)
> 	- Whether the longitude or latitude comes first in the pair. 
> If you copy coordinates from an attached media, you can also check a box to assign the media as a marker image. 
> - **Update markers**: Markers can have [customized thumbnails and labels](#editing-markers) rather than displaying just the item title as a link. You can batch-edit marker customization by:
> 	- Removing images
> 	- Using items' primary media as images
> 	- Copying labels from an existing metadata value (either from the item, the primary media, or the media already assigned to the marker)
> 	- Removing labels (the first entry in the dropdown).
> 
> ![The map-specific batch-editing field, filled out to copy "Latitude/Longitude" values from dcterms:coverage of each item.](modulesfiles/Mapping_batchEdit.png)
> 
> You may wish to copy your coordinates in one batch operation, then update those markers with labels and images in a second batch operation. Note that if an item has multiple markers, all markers will get a marker update of identical images and labels.

**Comment by allanaaa** [comment]
You can't batch-edit default bounds, right?

**Comment by allanaaa** [comment]
I'd like to do a bit more testing and documentation after this branch is merged into master, before the next release. Should only be a few days, but I want to make sure I have all the new bits sorted out and in the docs. Thanks!

**Comment by jimsafley** [comment]
I forgot to mention up top that this process will not create a marker if it has "identical" coordinates to an existing marker. By "identical" I mean a difference of 11-44 mm. At that precision there is no noticeable difference between coordinates on map on maximum zoom. It's probably worthwhile to include this information in the docs.

> You can't batch-edit default bounds, right?

That's correct. I'm going to wait on default bounds because I need to move on to other things.

**Comment by jimsafley** [comment]
I just merged the branch into master.
